python/delta-kernel-rust-sharing-wrapper/src/lib.rs

```diff
@@ -87,38 +123,67 @@ impl Scan {
         &self,
         engine_interface: &PythonInterface,
     ) -> DeltaPyResult<PyArrowType<Box<dyn RecordBatchReader + Send>>> {
-        let result_schema: SchemaRef =
-            Arc::new(self.0.schema().as_ref().try_into().map_err(|e| {
-                delta_kernel::Error::Generic(format!("Could not get result schema: {e}"))
-            })?);
-        let results = self.0.execute(engine_interface.0.as_ref())?;
-        let record_batches: Vec<_> = results
-            .map(|res| {
-                let scan_res = res.and_then(|res| Ok((res.full_mask(), res.raw_data?)));
-                let (mask, data) =
-                    scan_res.map_err(|e| ArrowError::from_external_error(Box::new(e)))?;
-                let record_batch: RecordBatch = data
-                    .into_any()
-                    .downcast::<ArrowEngineData>()
-                    .map_err(|_| {
-                        ArrowError::CastError("Couldn't cast to ArrowEngineData".to_string())
-                    })?
-                    .into();
-                if let Some(mask) = mask {
-                    let filtered_batch = filter_record_batch(&record_batch, &mask.into())?;
-                    Ok(filtered_batch)
-                } else {
-                    Ok(record_batch)
-                }
-            })
-            .collect();
-        let record_batch_iter = RecordBatchIterator::new(record_batches, result_schema);
+        let result_schema: ArrowSchemaRef = try_get_schema(self.0.schema())?;
+        let results = self.0.execute(engine_interface.0.clone())?;
+        let record_batch_iter = try_create_record_batch_iter(results, result_schema);
+        Ok(PyArrowType(Box::new(record_batch_iter)))
+    }
+}
+
+#[pyclass]
+struct TableChangesScanBuilder(Option<delta_kernel::table_changes::scan::TableChangesScanBuilder>);
+
+#[pymethods]
+impl TableChangesScanBuilder {
+    #[new]
+    #[pyo3(signature = (table, engine_interface, start_version, end_version=None))]
+    fn new(
+        table: &Table,
+        engine_interface: &PythonInterface,
+        start_version: u64,
+        end_version: Option<u64>,
+    ) -> DeltaPyResult<TableChangesScanBuilder> {
+        let table_changes =
+            table
+                .0
+                .table_changes(engine_interface.0.as_ref(), start_version, end_version)?;
+        Ok(TableChangesScanBuilder(Some(
+            table_changes.into_scan_builder(),
+        )))
+    }
+
+    fn build(&mut self) -> DeltaPyResult<TableChangesScan> {
+        let scan = self
+            .0
+            .take()
+            .ok_or_else(|| {
+                delta_kernel::Error::generic(
+                    "Can only call build() once on TableChangesScanBuilder",
+                )
+            })?
+            .build()?;
+        Ok(TableChangesScan(scan))
+    }
+}
```

Comment:
Currently we fail on any subsequent call to build. If we want to avoid the potential error, you can do something like this:
```suggestion
#[pyclass]
struct TableChangesScanBuilder {
    table_changes: Arc<TableChanges>,
}

#[pymethods]
impl TableChangesScanBuilder {
    #[new]
    #[pyo3(signature = (table, engine_interface, start_version, end_version=None))]
    fn new(
        table: &Table,
        engine_interface: &PythonInterface,
        start_version: u64,
        end_version: Option<u64>,
    ) -> DeltaPyResult<TableChangesScanBuilder> {
        let table_changes = table
            .0
            .table_changes(engine_interface.0.as_ref(), start_version, end_version)?
            .into();
        Ok(TableChangesScanBuilder { table_changes })
    }

    fn build(&mut self) -> DeltaPyResult<TableChangesScan> {
        // Note: cheap Arc clone
        let scan_builder = self.table_changes.clone().scan_builder();
        let scan = scan_builder.build()?;
        Ok(TableChangesScan(scan))
    }
}
```

Reply:
You can do the same thing in Snapsot/ScanBulider. @zachschuermann  would like to see what you think.

Reply:
I think if we change this behavior, we should change it for both TableChangesScanBuilder and ScanBuilder to be consistent.